src/controller/CommissionerDiscoveryController.cpp

```diff
@@ -121,86 +140,231 @@ void CallbackOk(System::Layer * aSystemLayer, void * aAppState)
 void CommissionerDiscoveryController::Ok()
 {
     ChipLogDetail(AppServer, "UX Ok: moving to main thread");
-    // need to ensure callback is on main chip thread
     assertChipStackLockedByCurrentThread();
-    DeviceLayer::SystemLayer().StartTimer(System::Clock::Seconds32(0), CallbackOk, this);
+    if (CHIP_NO_ERROR != DeviceLayer::SystemLayer().StartTimer(System::Clock::Seconds32(0), CallbackOk, this))
+    {
+        ChipLogError(AppServer, "UX Ok: StartTimer returned error");
+    }
 }
 
 void CommissionerDiscoveryController::InternalOk()
 {
     ChipLogDetail(AppServer, "UX InternalOk");
+    assertChipStackLockedByCurrentThread();
+    ValidateSession();
+
     if (!mPendingConsent)
     {
-        ChipLogError(AppServer, "UX Ok: no current instance");
+        ChipLogError(AppServer, "UX InternalOk: no current instance");
         return;
     }
     if (mUdcServer == nullptr)
     {
-        ChipLogError(AppServer, "UX Ok: no udc server");
+        ChipLogError(AppServer, "UX InternalOk: no udc server");
         return;
     }
     UDCClientState * client = mUdcServer->GetUDCClients().FindUDCClientState(mCurrentInstance);
     if (client == nullptr)
     {
-        ChipLogError(AppServer, "UX Ok: could not find instance=%s", mCurrentInstance);
+        ChipLogError(AppServer, "UX InternalOk: could not find instance=%s", mCurrentInstance);
         return;
     }
     if (client->GetUDCClientProcessingState() != UDCClientProcessingState::kPromptingUser)
     {
-        ChipLogError(AppServer, "UX Ok: invalid state for ok");
+        ChipLogError(AppServer, "UX InternalOk: invalid state for ok");
         return;
     }
     client->SetUDCClientProcessingState(UDCClientProcessingState::kObtainingOnboardingPayload);
 
-    if (mPasscodeService != nullptr)
+    if (mPasscodeService == nullptr)
     {
-        char rotatingIdString[chip::Dnssd::kMaxRotatingIdLen * 2 + 1] = "";
-        Encoding::BytesToUppercaseHexString(client->GetRotatingId(), client->GetRotatingIdLength(), rotatingIdString,
-                                            sizeof(rotatingIdString));
-        // Encoding::BytesToUppercaseHexString(client->GetRotatingId(), chip::Dnssd::kMaxRotatingIdLen, rotatingIdString,
-        //                                     sizeof(rotatingIdString));
+        HandleContentAppPasscodeResponse(0);
+        return;
+    }
 
-        CharSpan rotatingIdSpan = chip::CharSpan(rotatingIdString, sizeof(rotatingIdString));
-        uint32_t passcode       = 0;
-        uint8_t targetAppCount  = client->GetNumTargetAppInfos();
-        if (targetAppCount > 0)
+    char rotatingIdString[Dnssd::kMaxRotatingIdLen * 2 + 1] = "";
+    Encoding::BytesToUppercaseHexString(client->GetRotatingId(), client->GetRotatingIdLength(), rotatingIdString,
+                                        sizeof(rotatingIdString));
+    CharSpan rotatingIdSpan = CharSpan(rotatingIdString, strlen(rotatingIdString));
```

Comment:
If we actually want `strlen` here, we should be using `CharSpan::fromCharString`.

But what really confuses me is this: why are we doing any null-termination here at all?  Especially because there is absolutely 0 guarantee that this buffer is in fact null-terminated, because we are never checking for error from BytesToUppercaseHexString.

Why does this code not look like this:

```
char rotatingIdBuffer[Dnssd::kMaxRotatingIdLen * 2];
size_t rotatingIdLength = client->GetRotatingIdLength();
CHIP_ERROR err = Encoding::BytesToUppercaseHexBuffer(client->GetRotatingId(), rotatingIdLength, rotatingIdBuffer,
                                        sizeof(rotatingIdBuffer));
// Handle err being a failure here
CharSpan rotatingIdSpan(rotatingIdBuffer, 2*rotatingIdLength);
```

or so.

Reply:
fixed